module/zfs/dmu.c

```diff
@@ -1042,245 +1277,599 @@ dmu_read_by_dnode(dnode_t *dn, uint64_t offset, uint64_t size, void *buf,
- * (e.g. its bonus buffer), this routine is faster than dmu_read_uio(),
- * because we don't have to find the dnode_t for the object.
- */
-int
-dmu_read_uio_dbuf(dmu_buf_t *zdb, uio_t *uio, uint64_t size)
-{
-	dmu_buf_impl_t *db = (dmu_buf_impl_t *)zdb;
-	dnode_t *dn;
-	int err;
+static void
+dmu_write_direct_ready(zio_t *zio)
+{
+	dmu_sync_ready(zio, NULL, zio->io_private);
+}
+
+static void
+dmu_write_direct_done(zio_t *zio)
+{
+	dmu_sync_arg_t *dsa = zio->io_private;
+	dbuf_dirty_record_t *dr = dsa->dsa_dr;
+	dmu_buf_impl_t *db = dr->dr_dbuf;
+
+	if (abd_pages_are_stable(zio->io_abd))
+		abd_release_stable_pages(zio->io_abd);
+	abd_put(zio->io_abd);
+
+	mutex_enter(&db->db_mtx);
+	if (db->db_buf) {
+		arc_buf_t *buf = db->db_buf;
+		/*
+		 * The current contents of the dbuf are now stale.
+		 */
+		ASSERT(db->db_buf == dr->dt.dl.dr_data);
+		db->db_buf = NULL;
+		db->db.db_data = NULL;
+		dr->dt.dl.dr_data = NULL;
+		arc_buf_destroy(buf, db);
+	}
+	ASSERT(db->db.db_data == NULL);
+	db->db_state = DB_UNCACHED;
+	mutex_exit(&db->db_mtx);
+
+	dmu_sync_done(zio, NULL, zio->io_private);
+	kmem_free(zio->io_bp, sizeof (blkptr_t));
+}
+
+static boolean_t
+dmu_need_to_make_pages_stable(zio_prop_t zp)
+{
+	boolean_t make_stable = B_FALSE;
+	if (zp.zp_compress != ZIO_COMPRESS_OFF ||
+	    zp.zp_checksum != ZIO_CHECKSUM_OFF ||
+	    zp.zp_encrypt == B_TRUE)
+		make_stable = B_TRUE;
+	return (make_stable);
+}
+
+static int
+dmu_write_direct(zio_t *pio, dmu_buf_impl_t *db, abd_t *data, dmu_tx_t *tx,
+    boolean_t lustre_buf)
+{
+	objset_t *os = db->db_objset;
+	dsl_dataset_t *ds = os->os_dsl_dataset;
+	dbuf_dirty_record_t *dr_head, *dr_next;
+	dmu_sync_arg_t *dsa;
+	zbookmark_phys_t zb;
+	zio_prop_t zp;
+	dnode_t *dn;
+	uint64_t txg = dmu_tx_get_txg(tx);
+	blkptr_t *bp;
+	zio_t *zio;
+	int err = 0;
+
+	ASSERT(tx != NULL);
+
+	SET_BOOKMARK(&zb, ds->ds_object,
+	    db->db.db_object, db->db_level, db->db_blkid);
+
+	/*
+	 * No support for this
+	 */
+	if (txg > spa_freeze_txg(os->os_spa))
+		return (SET_ERROR(ENOTSUP));
+
+	DB_DNODE_ENTER(db);
+	dn = DB_DNODE(db);
+	dmu_write_policy(os, dn, db->db_level, WP_DMU_SYNC, &zp);
+	if (lustre_buf) {
+		/*
+		 * If the buffer we are writing is coming from Lustre we
+		 * do not have to worry about amking the pages stable as Lustre
+		 * has already made a stable copy of the data.
+		 */
+		zp.zp_lustre_buf = B_TRUE;
+	} else if (dmu_need_to_make_pages_stable(zp)) {
+		/*
+		 * If we are doing any of the following operations:
+		 * Compression, Checksum, Encryption, or Parity
+		 *
+		 * We have to make sure the pages are stable to prevent
+		 * the data from changing while any of those operations
+		 * are happening. If we did not do this, we could wind
+		 * up with user created checksum errors. At this point
+		 * we can not determine if parity is necessary but that
+		 * will be checked in the ZIO pipeline.
+		 */
+		zp.zp_lustre_buf = B_FALSE;
+		abd_make_pages_stable(data);
```

Comment:
Which I believe means at this point we'll always already have the pages held (via `iov_ter_get_pages_alloc()`) until the end of `zpl_{read,write}_iter` so there shouldn't be a need to pin them again.  Since pinning is essentially identical to the get with some slightly different reference counting.

Reply:
With the latest updates, we no longer have the `zp_lustre_buf` as all the pages are mapped in the ZPL layer.